### DISCUSSION TITLE
Move some aspects of `eq` from Model to Data

### DISCUSSION THREAD

**Discussion by PeterMitrano** [google-deepmind/mujoco#876] [Ideas]
I'm working on manipulation of deformable objects (rope) and I've found myself using `m.eq_active` and `m.eq_data` in ways that feel more like _data_ than they do _model_. Generally speaking, things in model should not change during the simulation. However, using equality constraints to simulate simplified grasping is very handy, and means you probably want to change `eq_data` and `eq_active` during the simulation. Right now if I want to save/restore/copy the simulation state I have to also remember to do those fields from `model` in addition to the entire `data` object. Might it make sense to move them to the MjData structure instead?

**Comment by yuvaltassa** [comment]
Yup this is a good idea that we've discussed before. However, removing things from `mjModel` is a bad idea in multiple ways.

My proposal would be to **add** `eq_active` and `eq_data` to `mjData`, initialised to the values in `mjModel`, which have the semantic of **overriding** the values in `mjModel`. So we keep the values in `mjModel`, but their semantic becomes "the values that the corresponding variables in `mjData` get upon `mj_resetData`". Does that sound reasonable?

**Reply by PeterMitrano** [reply]
Yes, that makes sense! Could the versions in the model be renamed to avoid confusing and signal that their meaning changes? possibly to `eq_data0` and `eq_active0`?

**Reply by yuvaltassa** [reply]
Technically, that is a breaking API change that we try very hard to avoid, but it is very consistent with the naming scheme. 

I think I'd support either your proposal of appending `0` to the `mjModel` variables, or naming the `mjData` variables `eq_data_override`. The latter is safer (in terms of API breakage), but uglier.

I'll try your proposal and see what ramifications it has in terms of code breakage...

**Reply by PeterMitrano** [reply]
Perhaps `obj2id` should also have an override in data? I'm less confident on that one, but for my application I do actually change which body the `eq` is between (gripper and then a changing body within the composite rope/cable object). You could also implement that with N different equality constraints, but that's a bit less elegant IMHO. So in some ways the object IDs is sometimes treated as a data variable :shrug:  potentially a slippery slope.

A main reason I'm concerned about the model/data distinction is that copying the MjModel object is ~10x slower than copy the MjData object. I'm not entirely sure why, but if MjModel were faster to copy, I wouldn't be as worried about what lives where.

**Reply by yuvaltassa** [reply]
Try removing visual only assets for smaller mjModels. We plan to introduce an option to loadXML that will strip visual assets. BTW, why does changing these numbers imply mjModel copying?

Regarding the OP, I think I've changed my mind (again 🙂), we should not put things like this in mjData that are not proper state. A slippery slope, like you say. 

There is indeed a very good use case for this: equality constraints that break or deform. You should be able to tell eq constraints that they break at some maximum force/torque or that weld constraints can be plastic, changing their relative base pose under stress. But these would be special constraints, behaving like actuators that are allowed to declare that they [carry state variables](https://mujoco.readthedocs.io/en/latest/computation.html#activation-dynamics). Then you actually have dynamics, so these would be mjData state variables proper. But it shouldn't be _all_ equality constraints. What you are doing is "magic" not physics, you are actually changing the model according to your "game logic", so modifying mjModel is correct.

1. Does this make sense?
2. If we had such breakable/bendable constraints, would you use them?

**Reply by PeterMitrano** [reply]
1. It's reasonable, and if copying the MjModel was faster I wouldn't much care. Why? Imagine I want to run a planner that considers grasping different bodies on a rope. I need to give that planner the initial state of the world (e.g, `m, d`) and in order to simulate the eq changes it will modify `m`, meaning I need to either (1) give it a `copy()` of `m` which can is slow compared to copying `d` or I need to manually save/restore the specific fields `eq_data, eq_active, eq_obj2id` from `m` so that after planning when I go back to simulating, what the planning did doesn't affect my actual simulation. Hopefully that makes sense? I don't really have "visual only assets", although maybe I'm not interpreting that phrase correctly.

2. I don't think force-based breakable/bendable constraints are what I want in this example.

**Reply by yuvaltassa** [reply]
(modifying item numbering)

1. Yes, of course that makes sense. The single-model-multiple-data approach ceases to be viable, you now need multiple models.
2. By "visual only assets" I mean textures and non-colliding meshes. For example in the model described in [this comment](https://github.com/deepmind/mujoco/issues/867#issuecomment-1553799265), removing the textures reduces `mjModel` size from 800KB to 5KB.
3. Okay, new proposal below.

- The breakable/bendable thing should be implemented in the spirit of actuators and is the only thing worthy of being called a "state variable".
- You want a different thing, you're asking for [user input variables](https://mujoco.readthedocs.io/en/latest/computation.html#the-state) like `ctrl` and `xfrc_applied`. I feel much more comfortable with this idea. How would you feel about something like

```xml
  <equality>
    <weld ... usersettable="true"/>
  <equality>
```

and 

```C
struct mjModel_ {
  ...
  int neq;                    // number of equality constraints
  int neq_user;               // number of user-settable equality constraints
  ...
};

...

struct mjData_ {
  ...
  // control
  mjtNum* ctrl;              // control                                          (nu x 1)
  mjtNum* qfrc_applied;      // applied generalized force                        (nv x 1)
  mjtNum* xfrc_applied;      // applied Cartesian force/torque                   (nbody x 6)
  mjtNum* eq_data_user;      // user numeric data for constraint                 (neq_user x mjNEQDATA)
  ...
};
```

I don't love "usersettable" or `_user` TBH. Maybe "override" for both? How about "input"?

**Reply by saran-t** [reply]
I prefer `<weld ... dynamic="true"/>`, `neqdyn` in `mjModel`, and `eq` in `mjData`.

**Reply by yuvaltassa** [reply]
@saran-t, that's exactly what we'll want for the dynamic constraints I describe in the first bullet above. What the OP is asking for is different, not "dynamic" like `mjData.act`, but user-settable inputs, like `mjData.ctrl`.

I'm leaning towards `<weld input="true">`, `mjModel.neq_input` and `mjData.eq_data_input`.

We'll probably need a new enum for equalities that means fixed/stateful/settable.

**Reply by saran-t** [reply]
"controllable"? or is that confusable with the mathematical notion of controllability?

**Reply by yuvaltassa** [reply]
Ya control implies applying forces which is not what this is. Also the reason it has to be `eq_data_input`. You can't delete the `_data` since we're not touching the constraint forces, just parameters. `_prm` would have been better than `_data` but too late now I guess.

The more I roll with `_input` the more I like it. Override is more precise but long and clunky.

Another thought I'm having is that `eq_active` should be a float. 0 means the same thing, remove from `efc`, but in (0, 1] it weakens the constraint force proportionally. Maybe @emotodorov knows of a principled way of doing this.

**Reply by PeterMitrano** [reply]
This seems like a good way to achieve what I want! It makes sense as a user input to physics like `ctrl`. Should `eq_active` also be a user input? So both `eq_data_user` and `eq_active_user`?

Thanks for pointing me to that section of the docs, I just learned a few things about what components of data I actually need to be copying, which could say a tiny bit of time :thinking: . I can also try stripping textures and see how much time that saves!

If `active` is a float, I think that would conflict with `solref`? for example `solref=0.02 1` vs `solref=0.04 1` where the latter is softer?

**Comment by emotodorov** [comment]
I think we should add eq_enable to mjData, initialize it to
mjModel.eq_enable in mj_reset(), and then use it at runtime to
enable/disable constraints.  Making it float does not work, since the
constraint forces are computed by the solver and it is not a good idea to
scale them afterwards.

Incidentally, in the new soft bodies I am adding, it also makes sense to do
this in mjData so as to cut/tear soft bodies.  I will have to add a similar
field for collision elements  in mjData (something like elem_enable) so
that both collisions and equality constrains can be disabled at runtime.
In that case though, there is no corresponding field in mjModel.

-- Emo

**Comment by yuvaltassa** [comment]
Fixed by ee78b8f76ba20b3d414a8ea1cda2cbe6e9cc96b6
